File: frontend/src/scripts/bewegungen.ts
/**
 * Das Bewegungsbild: welcher Ort wann von wem belegt ist, und wer zwischen welchen Orten
 * unterwegs ist.
 *
 * Jeder Ort ist ein waagerechtes Band, die Zeit läuft nach rechts. Ein Aufenthalt ist ein Balken
 * im Band, eine Fahrt eine Linie, die von einem Band ins andere zieht. Damit ist die Bewegung
 * selbst das Bild und nicht bloß eine Zeile in einer Tabelle.
 *
 * Erzählt wird der Tag wahlweise je Fahrzeug oder je Person. Beides ist dieselbe Rechnung über
 * **Spuren** — eine Spur ist eine Folge von Schritten mit einem Namen. Im Fahrzeugbild ist das
 * eine Kette, im Personenbild der Plan einer Person, der ohnehin aus denselben Ketten entsteht.
 *
 * Gerechnet wird hier nur die Anordnung, nicht das Aussehen: Bänder, Reihen darin, Zeitfenster.
 * Dieselbe Anordnung zeichnet der Bildschirm und das PDF, und `tools/plan_vergleichen` hält die
 * beiden Seiten aneinander.
 */
import type {Lauf, Schritt} from '../interfaces/Planung'
import type {Plandaten} from './ablauf'
import {personenplan, vonOrt} from './ablauf'
import {alsMinuten, tagVon} from './zeit'

/** Wovon der Tag erzählt wird. */
export type Modus = 'fahrzeuge' | 'personen'

/** Ein Ort als Band. `reihen` sagt, wie viele Balken darin nebeneinander liegen müssen. */
export interface Band {
    ortId: string
    name: string
    reihen: number
}

/** Ein Aufenthalt: jemand steht von — bis in diesem Band. */
export interface Balken {
    spurId: string
    schrittId: string
    ortId: string
    reihe: number
    /** Minuten seit Mitternacht des Tages; über Mitternacht hinaus größer als 1440. */
    von: number
    bis: number
    name: string
    /**
     * Wer oder was noch dazugehört: im Fahrzeugbild die Besatzung, im Personenbild das
     * Fahrzeug, in dem die Person sitzt. Leer, wo jemand für sich steht.
     */
    begleitung: string[]
    lage: string
}

/** Eine Fahrt: die Linie, die ein Band verlässt und in einem anderen ankommt. */
export interface Linie {
    spurId: string
    schrittId: string
    vonOrtId: string
    vonReihe: number
    nachOrtId: string
    nachReihe: number
    von: number
    bis: number
    mittel: Schritt['mittel']
    name: string
    begleitung: string[]
}

export interface Bewegungsbild {
    datum: string
    modus: Modus
    /** Das Zeitfenster, auf volle Stunden gelegt, damit die Achse runde Zahlen trägt. */
    von: number
    bis: number
    baender: Band[]
    balken: Balken[]
    linien: Linie[]
}

/** Ein Schritt, wie ihn eine Spur sieht: mit dem Ort, an dem er anfängt, und seinem Umfeld. */
interface Spurschritt {
    schritt: Schritt
    vonOrtId: string
    begleitung: string[]
}

interface Spur {
    id: string
    name: string
    schritte: Spurschritt[]
}

const STUNDE = 60

function minuteAmTag(zeitpunkt: string, datum: string): number {
    return (alsMinuten(zeitpunkt) ?? 0) - (alsMinuten(`${datum}T00:00`) ?? 0)
}

/** Die Tage, an denen überhaupt etwas geplant ist, in ihrer natürlichen Reihenfolge. */
export function bewegungstage(daten: Plandaten): string[] {
    const tage = new Set<string>()
    for (const lauf of daten.planung.laeufe) {
        for (const schritt of lauf.schritte) {
            const tag = tagVon(schritt.von)
            if (tag) tage.add(tag)
        }
    }
    return [...tage].sort()
}

function fahrzeugName(daten: Plandaten, fahrzeugId: string): string {
    return daten.fahrzeuge.find(fahrzeug => fahrzeug.id === fahrzeugId)?.funkrufname ?? ''
}

function personName(daten: Plandaten, personId: string): string {
    return daten.planung.personen.find(person => person.id === personId)?.name ?? ''
}

function laufName(daten: Plandaten, lauf: Lauf): string {
    return lauf.fahrzeugId
        ? fahrzeugName(daten, lauf.fahrzeugId)
        : personName(daten, lauf.personId)
}

function besatzung(daten: Plandaten, schritt: Schritt): string[] {
    return schritt.besatzung.map(platz => personName(daten, platz.personId))
}

/** Im Fahrzeugbild ist jede Kette eine Spur — die eines Fahrzeugs wie die einer Person. */
function fahrzeugspuren(daten: Plandaten): Spur[] {
    return daten.planung.laeufe.map(lauf => ({
        id: lauf.id, name: laufName(daten, lauf),
        schritte: lauf.schritte.map(schritt => ({
            schritt, vonOrtId: vonOrt(lauf, schritt), begleitung: besatzung(daten, schritt),
        })),
    }))
}

/**
 * Im Personenbild ist jede Person eine Spur. Ihr Plan wird nirgends gepflegt, sondern aus
 * denselben Ketten gerechnet — deshalb kann das Personenbild dem Fahrzeugbild nicht
 * widersprechen.
 */
function personenspuren(daten: Plandaten): Spur[] {
    return daten.planung.personen.map(person => ({
        id: person.id, name: person.name,
        schritte: personenplan(daten, person.id).map(eintrag => ({
            schritt: eintrag.schritt,
            vonOrtId: eintrag.vonOrtId,
            begleitung: eintrag.lauf.fahrzeugId
                ? [fahrzeugName(daten, eintrag.lauf.fahrzeugId)]
                : [],
        })),
    }))
}

/**
 * Die Reihe im Band: der erste Platz, der zu dieser Zeit frei ist. Zwei Fahrzeuge, die
 * nacheinander an demselben Ort stehen, teilen sich damit eine Reihe, und nur was wirklich
 * gleichzeitig dasteht, braucht Platz übereinander.
 */
function reiheSuchen(belegt: number[], von: number, bis: number): number {
    for (let reihe = 0; reihe < belegt.length; reihe++) {
        if (belegt[reihe]! <= von) {
            belegt[reihe] = bis
            return reihe
        }
    }
    belegt.push(bis)
    return belegt.length - 1
}

interface Anwesenheit {
    spur: Spur
    eintrag: Spurschritt
    von: number
    bis: number
}

/**
 * Das Bild eines Tages. Ein Ort bekommt ein Band, sobald jemand dort steht oder dorthin fährt;
 * Orte, an denen an diesem Tag nichts geschieht, tauchen nicht auf.
 */
export function bewegungsbild(daten: Plandaten, datum: string,
                              modus: Modus = 'fahrzeuge'): Bewegungsbild {
    const spuren = modus === 'personen' ? personenspuren(daten) : fahrzeugspuren(daten)
    const amTag = (eintrag: Spurschritt) => tagVon(eintrag.schritt.von) === datum

    const stehend: Anwesenheit[] = []
    for (const spur of spuren) {
        for (const eintrag of spur.schritte) {
            if (eintrag.schritt.art !== 'aufenthalt' || !amTag(eintrag)) continue
            stehend.push({spur, eintrag,
                          von: minuteAmTag(eintrag.schritt.von, datum),
                          bis: minuteAmTag(eintrag.schritt.bis, datum)})
        }
    }
    stehend.sort((a, b) => a.von - b.von)

    const fahrten = spuren.flatMap(spur => spur.schritte
        .filter(eintrag => eintrag.schritt.art === 'fahrt' && amTag(eintrag))
        .map(eintrag => ({spur, eintrag})))

    const beteiligt = new Set<string>()
    for (const {eintrag} of stehend) beteiligt.add(eintrag.schritt.ortId)
    for (const {eintrag} of fahrten) {
        beteiligt.add(eintrag.vonOrtId)
        beteiligt.add(eintrag.schritt.ortId)
    }
    const reihenfolge = daten.planung.orte
        .filter(ort => ort.id && beteiligt.has(ort.id))
        .map(ort => ort.id)

    const belegung = new Map<string, number[]>(reihenfolge.map(ortId => [ortId, []]))
    const reihen = new Map<string, number>()
    const balken: Balken[] = []
    for (const anwesend of stehend) {
        const belegt = belegung.get(anwesend.eintrag.schritt.ortId)
        if (!belegt) continue
        const reihe = reiheSuchen(belegt, anwesend.von, anwesend.bis)
        reihen.set(`${anwesend.spur.id}|${anwesend.eintrag.schritt.id}`, reihe)
        balken.push({
            spurId: anwesend.spur.id, schrittId: anwesend.eintrag.schritt.id,
            ortId: anwesend.eintrag.schritt.ortId, reihe,
            von: anwesend.von, bis: anwesend.bis, name: anwesend.spur.name,
            begleitung: anwesend.eintrag.begleitung,
            lage: daten.planung.programmpunkte
                .find(punkt => punkt.id === anwesend.eintrag.schritt.programmpunktId)?.name ?? '',
        })
    }

    const linien: Linie[] = fahrten.map(({spur, eintrag}) => {
        const stelle = spur.schritte.indexOf(eintrag)
        const abfahrt = spur.schritte[stelle - 1]
        const ankunft = spur.schritte[stelle + 1]
        const reihe = (nachbar: Spurschritt | undefined) =>
            nachbar ? reihen.get(`${spur.id}|${nachbar.schritt.id}`) ?? 0 : 0
        return {
            spurId: spur.id, schrittId: eintrag.schritt.id,
            vonOrtId: eintrag.vonOrtId, vonReihe: reihe(abfahrt),
            nachOrtId: eintrag.schritt.ortId, nachReihe: reihe(ankunft),
            von: minuteAmTag(eintrag.schritt.von, datum),
            bis: minuteAmTag(eintrag.schritt.bis, datum),
            mittel: eintrag.schritt.mittel, name: spur.name,
            begleitung: eintrag.begleitung,
        }
    })

    const zeiten = [...balken, ...linien].flatMap(eintrag => [eintrag.von, eintrag.bis])
    const von = zeiten.length ? Math.floor(Math.min(...zeiten) / STUNDE) * STUNDE : 0
    const bis = zeiten.length ? Math.ceil(Math.max(...zeiten) / STUNDE) * STUNDE : STUNDE
    return {
        datum, modus, von, bis: Math.max(bis, von + STUNDE),
        baender: reihenfolge.map(ortId => ({
            ortId,
            name: daten.planung.orte.find(ort => ort.id === ortId)?.name ?? '',
            reihen: Math.max(1, belegung.get(ortId)?.length ?? 1),
        })),
        balken, linien,
    }
}

/** Wo eine Kette zu einem Zeitpunkt ist: an einem Ort, oder zwischen zweien unterwegs. */
export interface Standort {
    laufId: string
    name: string
    art: 'fahrzeug' | 'person'
    /** Der Ort, an dem sie steht. Leer, solange sie unterwegs ist. */
    ortId: string
    unterwegs: Unterwegs | null
    personen: string[]
    lage: string
}

export interface Unterwegs {
    vonOrtId: string
    nachOrtId: string
    /** Wie weit die Strecke zurückgelegt ist, zwischen 0 und 1. */
    anteil: number
    mittel: Schritt['mittel']
}

/**
 * Der Stand zu einem Zeitpunkt — das, was am Ausführungstag zählt: wer steht wo, wer ist
 * unterwegs und wie weit. Wer zu dieser Zeit nichts geplant hat, taucht nicht auf.
 */
export function standorte(daten: Plandaten, zeitpunkt: string): Standort[] {
    const jetzt = alsMinuten(zeitpunkt)
    if (jetzt === null) return []
    const gefunden: Standort[] = []
    for (const lauf of daten.planung.laeufe) {
        const schritt = lauf.schritte.find(kandidat => {
            const von = alsMinuten(kandidat.von), bis = alsMinuten(kandidat.bis)
            return von !== null && bis !== null && von <= jetzt && jetzt < bis
        })
        if (!schritt) continue
        const von = alsMinuten(schritt.von) ?? 0
        const bis = alsMinuten(schritt.bis) ?? von + 1
        gefunden.push({
            laufId: lauf.id, name: laufName(daten, lauf),
            art: lauf.fahrzeugId ? 'fahrzeug' : 'person',
            ortId: schritt.art === 'fahrt' ? '' : schritt.ortId,
            unterwegs: schritt.art !== 'fahrt' ? null : {
                vonOrtId: vonOrt(lauf, schritt), nachOrtId: schritt.ortId,
                anteil: bis > von ? (jetzt - von) / (bis - von) : 0,
                mittel: schritt.mittel,
            },
            personen: lauf.personId
                ? [laufName(daten, lauf), ...besatzung(daten, schritt)]
                : besatzung(daten, schritt),
            lage: daten.planung.programmpunkte
                .find(punkt => punkt.id === schritt.programmpunktId)?.name ?? '',
        })
    }
    return gefunden
}

/** Was als Nächstes ansteht. */
export interface Ereignis {
    zeitpunkt: string
    art: 'abfahrt' | 'ankunft'
    name: string
    ortId: string
    lage: string
    /** Minuten von jetzt bis dahin. */
    in: number
}

/**
 * Die nächsten Ereignisse ab einem Zeitpunkt: jede Abfahrt und jede Ankunft, in der Reihenfolge,
 * in der sie eintreten. Am Ausführungstag ist das die eigentliche Frage — was kommt als
 * Nächstes, und wie lange noch.
 */
export function ereignisse(daten: Plandaten, zeitpunkt: string, anzahl = 8): Ereignis[] {
    const jetzt = alsMinuten(zeitpunkt)
    if (jetzt === null) return []
    const gefunden: Ereignis[] = []
    for (const lauf of daten.planung.laeufe) {
        for (const schritt of lauf.schritte) {
            const beginn = alsMinuten(schritt.von)
            if (beginn === null || beginn < jetzt) continue
            gefunden.push({
                zeitpunkt: schritt.von,
                art: schritt.art === 'fahrt' ? 'abfahrt' : 'ankunft',
                name: laufName(daten, lauf), ortId: schritt.ortId,
                lage: daten.planung.programmpunkte
                    .find(punkt => punkt.id === schritt.programmpunktId)?.name ?? '',
                in: beginn - jetzt,
            })
        }
    }
    return gefunden.sort((a, b) => a.in - b.in).slice(0, anzahl)
}

File: backend/src/data/bewegungen.py
"""
Das Bewegungsbild für den Druck.

Jeder Ort ist ein waagerechtes Band, die Zeit läuft nach rechts, ein Aufenthalt ist ein Balken
darin und eine Fahrt eine Linie von einem Band ins andere. Erzählt wird der Tag wahlweise je
Fahrzeug oder je Person; beides ist dieselbe Rechnung über **Spuren** — eine Spur ist eine Folge
von Schritten mit einem Namen. Im Fahrzeugbild ist das eine Kette, im Personenbild der Plan einer
Person, der ohnehin aus denselben Ketten entsteht.

Dieselbe Anordnung rechnet der Browser in `frontend/src/scripts/bewegungen.ts` für den
Bildschirm. Die beiden müssen übereinstimmen, sonst zeigt der Ausdruck ein anderes Bild als die
Ansicht; `tools/plan_vergleichen` hält sie aneinander.
"""

from dataclasses import dataclass, field

from data.planzeit import minuten as _minuten, tag as _tag
from entities.alarm import Arbeitsmappe
from entities.planung import Lauf, Person, Planung, Schritt

STUNDE = 60

MODI = ("fahrzeuge", "personen")


def _minute_am_tag(zeitpunkt: str, datum: str) -> int:
    return (_minuten(zeitpunkt) or 0) - (_minuten(f"{datum}T00:00") or 0)


def _von_ort(lauf: Lauf, schritt: Schritt) -> str:
    """Wo eine Fahrt losgeht, sagt der vorige Schritt."""
    if schritt.art != "fahrt":
        return schritt.ortId
    stelle = lauf.schritte.index(schritt)
    return lauf.schritte[stelle - 1].ortId if stelle > 0 else schritt.ortId


def bewegungstage(planung: Planung) -> list[str]:
    """Die Tage, an denen überhaupt etwas geplant ist."""
    tage = {_tag(schritt.von) for lauf in planung.laeufe for schritt in lauf.schritte}
    return sorted(tag for tag in tage if tag)


@dataclass
class _Spurschritt:
    """Ein Schritt, wie ihn eine Spur sieht: mit dem Ort, an dem er anfängt, und seinem Umfeld."""

    schritt: Schritt
    von_ort_id: str
    begleitung: list[str]


@dataclass
class _Spur:
    id: str
    name: str
    schritte: list[_Spurschritt] = field(default_factory=list)


def _reihe_suchen(belegt: list[int], von: int, bis: int) -> int:
    """
    Der erste Platz, der zu dieser Zeit frei ist. Zwei Fahrzeuge, die nacheinander am selben Ort
    stehen, teilen sich damit eine Reihe; nur Gleichzeitiges braucht Platz übereinander.
    """
    for reihe, ende in enumerate(belegt):
        if ende <= von:
            belegt[reihe] = bis
            return reihe
    belegt.append(bis)
    return len(belegt) - 1


class _Namen:
    def __init__(self, arbeitsmappe: Arbeitsmappe) -> None:
        self.planung = arbeitsmappe.planung
        self.orte = arbeitsmappe.kataloge.alle_orte()
        self._orte = {ort.id: ort.name for ort in self.orte}
        self._personen = {person.id: person for person in self.planung.personen}
        self._lagen = {punkt.id: punkt.name for punkt in self.planung.programmpunkte}
        self._fahrzeuge = {
            fahrzeug.id: fahrzeug.funkrufname for fahrzeug in arbeitsmappe.kataloge.fahrzeuge}

    def ort(self, ort_id: str) -> str:
        return self._orte.get(ort_id, "")

    def lage(self, schritt: Schritt) -> str:
        return self._lagen.get(schritt.programmpunktId, "")

    def fahrzeug(self, fahrzeug_id: str) -> str:
        return self._fahrzeuge.get(fahrzeug_id, "")

    def person(self, person_id: str) -> str:
        eintrag = self._personen.get(person_id)
        return eintrag.name if eintrag else ""

    def kette(self, lauf: Lauf) -> str:
        return self.fahrzeug(lauf.fahrzeugId) if lauf.fahrzeugId else self.person(lauf.personId)

    def besatzung(self, schritt: Schritt) -> list[str]:
        return [self.person(platz.personId) for platz in schritt.besatzung]


def _fahrzeugspuren(namen: _Namen) -> list[_Spur]:
    """Im Fahrzeugbild ist jede Kette eine Spur — die eines Fahrzeugs wie die einer Person."""
    return [
        _Spur(lauf.id, namen.kette(lauf), [
            _Spurschritt(schritt, _von_ort(lauf, schritt), namen.besatzung(schritt))
            for schritt in lauf.schritte])
        for lauf in namen.planung.laeufe
    ]


def _personenspur(namen: _Namen, person: Person) -> _Spur:
    """
    Der Plan einer Person: alle Schritte, in deren Besatzung sie steht, plus die ihrer eigenen
    Kette, nach Zeit sortiert. Er wird nirgends gepflegt, sondern hieraus gelesen — deshalb kann
    das Personenbild dem Fahrzeugbild nicht widersprechen.
    """
    eintraege = []
    for lauf in namen.planung.laeufe:
        eigene = lauf.personId == person.id
        for schritt in lauf.schritte:
            sitzt = any(platz.personId == person.id for platz in schritt.besatzung)
            if not eigene and not sitzt:
                continue
            begleitung = [namen.fahrzeug(lauf.fahrzeugId)] if lauf.fahrzeugId else []
            eintraege.append((_minuten(schritt.von) or 0,
                              _Spurschritt(schritt, _von_ort(lauf, schritt), begleitung)))
    eintraege.sort(key=lambda eintrag: eintrag[0])
    return _Spur(person.id, person.name, [eintrag for _, eintrag in eintraege])


def bewegungsbild(arbeitsmappe: Arbeitsmappe, datum: str, modus: str = "fahrzeuge") -> dict:
    """
    Das Bild eines Tages. Ein Ort bekommt ein Band, sobald jemand dort steht oder dorthin fährt;
    wo an diesem Tag nichts geschieht, gibt es auch kein Band.
    """
    namen = _Namen(arbeitsmappe)
    spuren = ([_personenspur(namen, person) for person in namen.planung.personen]
              if modus == "personen" else _fahrzeugspuren(namen))

    stehend = sorted(
        ((spur, eintrag,
          _minute_am_tag(eintrag.schritt.von, datum), _minute_am_tag(eintrag.schritt.bis, datum))
         for spur in spuren for eintrag in spur.schritte
         if eintrag.schritt.art == "aufenthalt" and _tag(eintrag.schritt.von) == datum),
        key=lambda anwesend: anwesend[2])
    fahrten = [(spur, eintrag) for spur in spuren for eintrag in spur.schritte
               if eintrag.schritt.art == "fahrt" and _tag(eintrag.schritt.von) == datum]

    beteiligt = {eintrag.schritt.ortId for _, eintrag, _, _ in stehend}
    for _, eintrag in fahrten:
        beteiligt |= {eintrag.von_ort_id, eintrag.schritt.ortId}
    reihenfolge = [ort.id for ort in namen.orte if ort.id and ort.id in beteiligt]

    belegung: dict[str, list[int]] = {ort_id: [] for ort_id in reihenfolge}
    reihen: dict[tuple[str, str], int] = {}
    balken = []
    for spur, eintrag, von, bis in stehend:
        if eintrag.schritt.ortId not in belegung:
            continue
        reihe = _reihe_suchen(belegung[eintrag.schritt.ortId], von, bis)
        reihen[(spur.id, eintrag.schritt.id)] = reihe
        balken.append({
            "spurId": spur.id, "schrittId": eintrag.schritt.id,
            "ortId": eintrag.schritt.ortId, "reihe": reihe, "von": von, "bis": bis,
            "name": spur.name, "begleitung": eintrag.begleitung,
            "lage": namen.lage(eintrag.schritt),
        })

    linien = []
    for spur, eintrag in fahrten:
        stelle = spur.schritte.index(eintrag)
        abfahrt = spur.schritte[stelle - 1] if stelle > 0 else None
        ankunft = spur.schritte[stelle + 1] if stelle + 1 < len(spur.schritte) else None
        linien.append({
            "spurId": spur.id, "schrittId": eintrag.schritt.id,
            "vonOrtId": eintrag.von_ort_id,
            "vonReihe": reihen.get((spur.id, abfahrt.schritt.id), 0) if abfahrt else 0,
            "nachOrtId": eintrag.schritt.ortId,
            "nachReihe": reihen.get((spur.id, ankunft.schritt.id), 0) if ankunft else 0,
            "von": _minute_am_tag(eintrag.schritt.von, datum),
            "bis": _minute_am_tag(eintrag.schritt.bis, datum),
            "mittel": eintrag.schritt.mittel, "name": spur.name,
            "begleitung": eintrag.begleitung,
        })

    zeiten = [wert for eintrag in balken + linien for wert in (eintrag["von"], eintrag["bis"])]
    von = min(zeiten) // STUNDE * STUNDE if zeiten else 0
    bis = -(-max(zeiten) // STUNDE) * STUNDE if zeiten else STUNDE
    return {
        "datum": datum, "modus": modus, "von": von, "bis": max(bis, von + STUNDE),
        "baender": [{"ortId": ort_id, "name": namen.ort(ort_id),
                     "reihen": max(1, len(belegung[ort_id]))}
                    for ort_id in reihenfolge],
        "balken": balken, "linien": linien,
    }


def bewegungsbilder(arbeitsmappe: Arbeitsmappe) -> list[dict]:
    """Beide Erzählweisen, jede für jeden Tag — der Bogen zeigt sie nacheinander."""
    return [bewegungsbild(arbeitsmappe, tag, modus)
            for modus in MODI
            for tag in bewegungstage(arbeitsmappe.planung)]
